File: prob21.ts
//   Definition for singly-linked list.
class ListNode {
  val: number | null;
  next: ListNode | null;
  constructor(val?: number | null, next?: ListNode | null) {
    this.val = val === undefined ? null : val;
    this.next = next === undefined ? null : next;
  }
}

function mergeTwoLists(
  list1: ListNode | null,
  list2: ListNode | null
): ListNode | null {
  let pointer1 = list1;
  let pointer2 = list2;
  let mergedHead: ListNode | null = null,
    mergedNode: ListNode | null,
    prevMergedNode: ListNode | null = null;
  //The while loop stops when both pointer1 and pointer2 is null
  while (pointer1 || pointer2) {
    if (
      pointer1?.val !== undefined &&
      pointer1?.val !== null &&
      pointer2?.val !== undefined &&
      pointer2?.val !== null
    ) {
      if (pointer1?.val > pointer2?.val) {
        mergedNode = new ListNode(pointer2.val);
        pointer2 = pointer2.next;
        if (!mergedHead) {
          mergedHead = mergedNode;
        }
      } else {
        mergedNode = new ListNode(pointer1.val);
        pointer1 = pointer1?.next;
        if (!mergedHead) {
          mergedHead = mergedNode;
        }
      }
    } else if (pointer1?.val !== undefined && pointer1?.val !== null) {
      mergedNode = new ListNode(pointer1.val);
      pointer1 = pointer1?.next;
      if (!mergedHead) {
        mergedHead = mergedNode;
      }
    } else if (pointer2?.val !== undefined && pointer2?.val !== null) {
      mergedNode = new ListNode(pointer2?.val);
      pointer2 = pointer2.next;
      if (!mergedHead) {
        mergedHead = mergedNode;
      }
    } else {
      break;
    }
    if (prevMergedNode) prevMergedNode.next = mergedNode;
    prevMergedNode = mergedNode;
  }
  return mergedHead;
}

File: prob21.py
# Definition for singly-linked list.
from typing import Optional
import unittest

class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

class Solution:
    def mergeTwoLists(self, list1: Optional[ListNode], list2: Optional[ListNode]) -> Optional[ListNode]:
        prev = None
        ans= None

        while list1 or list2 :
            val1 = list1.val if list1 else list2.val+1
            val2 = list2.val if list2 else list1.val+1
            if val2<val1:
                curr_val = val2
                list2 = list2.next
            else:
                curr_val = val1
                list1 = list1.next
            if prev:
                prev.next = ListNode(curr_val)
                prev = prev.next
            else:
                prev = ListNode(curr_val)
                ans = prev

        return ans

class TestMergeTwoSortedLists(unittest.TestCase):
    def setUp(self):
        self.solution = Solution()

    def list_to_linked_list(self, values):
        if not values:
            return None
        head = ListNode(values[0])
        current = head
        for val in values[1:]:
            current.next = ListNode(val)
            current = current.next
        return head

    def linked_list_to_list(self, node):
        result = []
        while node:
            result.append(node.val)
            node = node.next
        return result

    def test_merge_two_sorted_lists_basic(self):
        list1 = self.list_to_linked_list([1, 3, 5])
        list2 = self.list_to_linked_list([2, 4, 6])
        expected_output = [1, 2, 3, 4, 5, 6]
        merged_list = self.solution.mergeTwoLists(list1, list2)
        self.assertEqual(self.linked_list_to_list(merged_list), expected_output)

    # def test_merge_two_sorted_lists_with_duplicates(self):
    #     list1 = self.list_to_linked_list([1, 2, 2, 3])
    #     list2 = self.list_to_linked_list([2, 2, 4, 5])
    #     expected_output = [1, 2, 2, 2, 2, 3, 4, 5]
    #     merged_list = self.solution.mergeTwoLists(list1, list2)
    #     self.assertEqual(self.linked_list_to_list(merged_list), expected_output)

if __name__ == "__main__":
    unittest.main()